### PR TITLE
feat(spanner): Add x-goog-spanner-end-to-end-tracing header for requests to Spanner

### PR DESCRIPTION
**Pull request by nareshz**
To enable tracing on Spanner layer, client can opt-in by following ways:

1. Set the `EnableEndToEndTracing` flag in ClientConfig to `true`. 
2. Set the `SPANNER_ENABLE_END_TO_END_TRACING` environment variable to `true`.

If enabled a new http header `x-goog-spanner-end-to-end-tracing` will be passed in Spanner requests.

**Note**: Spanner tracing also requires trace context propagation. This needs to be setup while setting OpenTelemetry.

Design doc: go/cs-tracing-mvp

### CODE REVIEW

File: spanner/client.go [hunk 1]
```diff
@@ -63,6 +63,10 @@ const (
 
 	requestsCompressionHeader = "x-response-encoding"
 
+	// endToEndTracingHeader is the name of the metadata header if client
+	// has opted-in for the creation of trace span's on the Spanner layer.
```
Comment (dsimil) [inline spanner/client.go:67]: nit: trace spans.

File: spanner/client.go [hunk 2]
```diff
@@ -331,6 +335,12 @@ type ClientConfig struct {
 	DirectedReadOptions *sppb.DirectedReadOptions
 
 	OpenTelemetryMeterProvider metric.MeterProvider
+
+	// EnableEndToEndTracing indicates whether end to end tracing is enabled or not.
+	// If it is enabled, trace span's will be created at Spanner layer.
```
Comment (dsimil) [inline spanner/client.go:340]: Same here.

**Review by dsimil — APPROVED** [review]

File: spanner/client.go (hunk 1 above)
Comment (nareshz) [reply spanner/client.go]: Done.

**Review by rahul2393 — APPROVED** [review]

**Review by harshachinta — APPROVED** [review]

**Comment by codyoss** [comment]
@rahul2393 can this be merged?

**Comment by harshachinta** [comment]
@dsimil Is this ready to be merged. I assume that there is testing going on in backend and this needs to wait until backend is ready?

File: spanner/client.go [hunk 3]
```diff
@@ -446,6 +458,19 @@ func newClientWithConfig(ctx context.Context, database string, config ClientConf
...
 		md.Append(requestsCompressionHeader, gzip.Name)
 	}
+	// Append server side tracing header if SPANNER_ENABLE_SERVER_SIDE_TRACING
+	// environment variable has been set or client has passed the opt-in option
+	// in ClientConfig.
+	appendServerSideTracingHeader := false
+	if config.EnableServerSideTracing {
+		appendServerSideTracingHeader = true
+	}
+	if serverSideTracingEnvironmentVariable := os.Getenv("SPANNER_ENABLE_SERVER_SIDE_TRACING"); serverSideTracingEnvironmentVariable == "true" {
```
Comment (nareshz) [inline spanner/client.go:468]: @harshachinta Added the option to enable server side using ENV variable. PTAL.

**Pull request merged**